lms/views/dashboard/api/course.py

```diff
@@ -70,18 +73,21 @@ def courses(self) -> APICourses:
     )
     def organization_courses(self) -> APICourses:
         filter_by_h_userids = self.request.parsed_params.get("h_userids")
+        filter_by_course_ids = self.request.parsed_params.get("course_ids")
         filter_by_h_assignment_ids = self.request.parsed_params.get("assignment_ids")
         org = self.dashboard_service.get_request_organization(self.request)
-        courses = self.request.db.scalars(
-            self.course_service.get_courses(
-                organization=org,
-                instructor_h_userid=self.request.user.h_userid
-                if self.request.user
-                else None,
-                h_userids=filter_by_h_userids,
-                assignment_ids=filter_by_h_assignment_ids,
-            )
-        ).all()
+        courses_query = self.course_service.get_courses(
+            organization=org,
+            instructor_h_userid=self.request.user.h_userid
+            if self.request.user
+            else None,
+            h_userids=filter_by_h_userids,
+            assignment_ids=filter_by_h_assignment_ids,
+        )
+        if filter_by_course_ids:
+            courses_query = courses_query.where(Course.id.in_(filter_by_course_ids))
```

Comment:
What is the policy for deciding whether to do the filtering inside `get_courses` or outside?

Reply:
I don't think there's clear policy.

The first and often best candidate is the service, we can abstract any complexity about models/queries there.

In this case I reckon the code reads better: we get a complex query for all the relevant rows from the service and filter it down immediately by a list of IDs from the parameter the view parameter. Those IDs are the PK of the table we are more interested, there's no joins or any extra complexity.

I reckon there's a risk of a future performance "improvement" if looking at the query in isolation in the service. When filtering a set of rows by the PK of in the same table we could say, just get the rows with this IDs and avoid doing all the other fluff, but that will be of course wrong.


That being said I don't have a strong opinion either way.